### ISSUE TITLE
Stratum v2 via IPC Mining Interface tracking issue

### ISSUE THREAD

**Issue by Sjors** [bitcoin/bitcoin#31098]
After much discussion in https://github.com/bitcoin/bitcoin/pull/29432 it appears there's currently little support for directly supporting Stratum v2 in Bitcoin Core, with most contributors favouring the creation of a Mining IPC interface that external applications can use.

The only such application currently is the Template Provider implemented by me in https://github.com/Sjors/bitcoin/pull/48. This also serves to illustrate how the interface is used and that it's complete. The folks working on SRI should be able to build a Template Provider as well, once the interface is complete and shipped in a release.

### Needed for (hopefully) v29

To complete the interface:

- [ ] #31003
- [ ] a few followups coming soon

For multiprocess release binaries

- [ ] #30983
- [ ] #30975 (one possible approach)
- [ ] https://github.com/chaincodelabs/libmultiprocess/pull/119
- [ ] https://github.com/chaincodelabs/libmultiprocess/issues/117

### Can wait for later releases

- [ ] https://github.com/chaincodelabs/libmultiprocess/issues/114
- [ ] #31002 DATUM developers are currently uninterested in taking advantage of the more performant interface, preferring to use the `getblocktemplate` RPC instead, despite an earlier concept ACK on Stratum v2 support https://github.com/bitcoin/bitcoin/pull/27854#issuecomment-1605237375.

### Current Template Provider users 

As a stop-gap measure I plan to maintain https://github.com/Sjors/bitcoin/pull/68 which is a continuation of https://github.com/bitcoin/bitcoin/pull/29432, since people are currently using that. Once the Mining interface makes it into a release, along with multiprocess enabled Guix binaries, I plan to drop that in favour of https://github.com/Sjors/bitcoin/pull/48.

Additionally I plan to convert the latter from a patch set on top of Bitcoin Core into a standalone application, to make it easier for others to contribute. That should make it less dependent on my own effort. The Bitcoin Core project could decide in the future to "adopt" this project, but that is entirely uncertain and most contributors seems averse to this idea.

**Comment by Sjors** [comment]
To address the last point (will respond to the rest later):

> It's also my understanding the the IPC logic in Bitcoin Core is not intended as a public API, and thus users of it are expected to exclusively be in Bitcoin Core, I suppose that is changing here?

In general IPC interfaces are not intended to be public, but the mining interface is an exception. @josibake is working on a proof-of-concept Rust demo that uses it, which should be fairly easy to use in SRI.

**Comment by Sjors** [comment]
> Do you think the lack of desire for review (compared to reviewing the capnproto protocol instead) comes from the encryption, the framing logic, the connman management, etc?

The "instead" misses the mark: reviewing capnproto was already necessary to achieve all the other use cases of multiprocess. The mining interface just provided an easier opportunity to make progress there.

Trying to redesign the interface in a way that makes it DoS-safe brings back the need for a lot of the review that _not_ including https://github.com/Sjors/bitcoin/pull/68 is trying to avoid. Which in practice probably means the interface won't get shipped for ages.

**Comment by plebhash** [comment]
on the pre-IPC Sv2 Template Provider that @Sjors provides [on his `sv2` branch](https://github.com/Sjors/bitcoin), we have two CLI parameters for `bitcoind`:

- `-sv2interval`
- `-sv2feedelta`

these two parameters dictated how often TP would send templates to its clients (Pool or JDC), which could happen either on a regular interval (`sv2interval`), or ad-hoc when needed (`sv2feedelta`).

it seems that the new IPC approach makes communication between `bitcoin-stratum` and `bitcoin-node` completely poll-based.

in other words, it's up to `bitcoin-stratum` to continuously poll `bitcoin-node` for new template changes.

wouldn't that potentially introduce inefficiencies? what if fee delta goes above the desired threshold? is `bitcoin-stratum` forced to wait until the next polling cycle before it get the update?

**Comment by Sjors** [comment]
@plebhash I pushed an update to the `bitcoin-mine` (Template Provider) application yesterday in https://github.com/Sjors/bitcoin/pull/48. It has the same `-sv2interval` and `-sv2feedelta` arguments. This uses `waitNext()` from #31283 under the hood. There's no polling involved!

**Comment by Sjors** [comment]
In both versions `-sv2interval` specifies the _minimum_ time before sending a fee update, and it currently can't set lower than 1 second. This minimum is not related to polling between two processes.

It's there because determining the new fees requires the node to construct a whole new block template under the hood. This is currently inefficient so the node waits one second between each template generation. But once it finds a better template the node _pushses_ it to the `bitcoin-mine` process.

The minimum interval can hopefully be dropped after Cluster Mempool is done #30289, as that should make it much cheaper to figure out how much fees for the next block have increased. At that point you can just set `-sv2interval=0` so it won't require an interface change.

The default value is 30 seconds which is probably a bit overkill conservative. It's there to avoid spamming the pool. Happy to take recommendations for a different default.

**Comment by ryanofsky** [comment]
I added #31375 to the list in the description since I think it will also be needed for a multirprocess release

**Comment by Sjors** [comment]
#31346 landed, marked as such in the description. Many to go, most of them are pretty simple and already have some review; just need more eyes to get them over the finish line.

**Comment by Sjors** [comment]
Updated description to hint at some new interface methods I plan to propose in the future: `createFutureBlock()`, `waitNewPowBlock()` and `verifyBlock()`.

However for now let's focus on the last interface method that the Template Provider needs, namely `waitNext()`. And on all the multiprocess changes.

**Comment by Sjors** [comment]
I moved Windows support to "can wait for later releases" and expanded the item a bit.

**Comment by Sjors** [comment]
The discussion in [todays IRC meeting](https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2025-01-23#1737649417-1737651814;) suggests v29 is very unlikely to happen, but v30 is conceivable.

**Comment by Sjors** [comment]
All interface changes that I need for v30 have been merged!

Some progress has been made on enabling IPC for the release as well, but please review #31375 and https://github.com/bitcoin/bitcoin/pull/31802.

**Comment by Sjors** [comment]
The wrapper binary PR #31375 just landed, which is great progress!

The only big change left is to add bitcoin-node (aka `bitcoin -m`) to the release binaries, which is what #31802 does.

**Comment by ryanofsky** [comment]
Does it make sense to close this issue?

**Comment by Sjors** [comment]
Yes, I might open a fresh tracking issue later for remaining and new interface changes. There are a few things that we want to adjust, including things that aren't urgent, but should be done when we need a breaking change anyway.

The sidecar application lives at https://github.com/Sjors/sv2-tp

**Comment by Sjors** [comment]
There's a fresh tracking issue for the mining IPC at #33777.
